rasa/core/policies/rule_policy.py

```diff
@@ -214,85 +310,125 @@ def predict_action_probabilities(
-
-                result[recalled] = 1
-            else:
-                logger.debug("There is no applicable rule.")
+        # there could be several negative rules
+        negative_recalled = [
+            self.lookup[NEGATIVE_RULES].get(key) for key in negative_keys
+        ]
 
-        return result
+        active_form_name = tracker.active_form_name()
+        if active_form_name:
+            # Check if a rule that predicted action_listen
+            # was applied inside the form.
+            # Rules might not explicitly switch back to the `Form`.
+            # Hence, we have to take care of that.
+            predicted_listen_from_general_rule = (
+                recalled == ACTION_LISTEN_NAME
+                and ACTIVE_FORM_PREFIX + active_form_name not in key
+            )
+            if predicted_listen_from_general_rule:
+                if NO_ACTIVE_FORM not in negative_recalled:
+                    # negative rules don't contain a key that corresponds to
+                    # the fact that active_form shouldn't be predicted
+                    logger.debug(
+                        f"Predicted form '{active_form_name}' by overwriting "
+                        f"'{ACTION_LISTEN_NAME}' predicted by general rule."
+                    )
+                    return active_form_name
+
+                # do not predict anything
+                recalled = None
+
+            # Since rule snippets and stories inside the form contain
+            # only unhappy paths, notify the form that
+            # it was predicted after an answer to a different question and
+            # therefore it should not validate user input for requested slot
+            if NO_VALIDATION in negative_recalled:
+                logger.debug("Added `FormValidation(False)` event.")
+                tracker.update(FormValidation(False))
+
+        if recalled is not None:
+            logger.debug(f"There is a rule for the next action '{recalled}'.")
+        else:
+            logger.debug("There is no applicable rule.")
+
+        return recalled
 
+    def predict_action_probabilities(
+        self,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter = RegexInterpreter(),
+        **kwargs: Any,
+    ) -> List[float]:
+        """Predicts the next action the bot should take after seeing the tracker.
 
-def _should_run_rasa_default_action(tracker: DialogueStateTracker) -> Optional[Text]:
-    if (
-        not tracker.latest_action_name == ACTION_LISTEN_NAME
-        or not tracker.latest_message
-    ):
-        return None
+        Returns the list of probabilities for the next actions.
+        If memorized action was found returns 1 for its index,
+        else returns 0 for all actions.
+        """
+        result = self._default_predictions(domain)
 
-    return DEFAULT_ACTION_MAPPINGS.get(tracker.latest_message.intent.get("name"))
+        if not self.is_enabled:
```

Comment:
can we throw out `toggle`  and `is_enabled`?

Reply:
I think we can, they are here for very long time and I don't know where they are used 🙈